MantidPlot/src/ProjectRecoveryThread.cpp

```diff
@@ -0,0 +1,326 @@
+
+#include "MantidAPI/FileProperty.h"
+#include "MantidKernel/ConfigService.h"
+#include "MantidKernel/Logger.h"
+
+#include "boost/optional.hpp"
+
+#include "Poco/DirectoryIterator.h"
+#include "Poco/NObserver.h"
+
+#include "Poco/Path.h"
+#include "qmetaobject.h"
+
+#include <chrono>
+#include <condition_variable>
+#include <ctime>
+#include <iomanip>
+#include <iostream>
+#include <mutex>
+#include <string>
+#include <thread>
+
+namespace {
+// Config helper methods
+
+template <typename T>
+boost::optional<T> getConfigValue(const std::string &key) {
+  T returnedValue;
+
+  int valueIsGood =
+      Mantid::Kernel::ConfigService::Instance().getValue<T>(key, returnedValue);
+
+  if (valueIsGood != 1) {
+    return boost::optional<T>{};
+  }
+
+  return boost::optional<T>{returnedValue};
+}
+
+boost::optional<bool> getConfigBool(const std::string &key) {
+  auto returnedValue = getConfigValue<std::string>(key);
+  if (!returnedValue.is_initialized()) {
+    return boost::optional<bool>{};
+  }
+
+  return returnedValue->find("true") != std::string::npos;
+}
+
+/// Returns a string to the current top level recovery folder
+std::string getRecoveryFolder() {
+  static std::string recoverFolder =
+      Mantid::Kernel::ConfigService::Instance().getAppDataDir() + "/recovery/";
+  return recoverFolder;
+}
+
+/// Gets a formatted timestamp
+std::string getTimeStamp() {
+  const char *formatSpecifier = "%Y-%m-%d %H-%M-%S";
+  auto time = std::time(nullptr);
+  auto localTime = std::localtime(&time);
+
+#if __GNUG__ < 5
+  // Have to workaround GCC 4 not having std::put_time on RHEL7
+  // this ifdef can be removed when RHEL7 uses a newer compiler
+  char timestamp[20];
+  if (strftime(timestamp, sizeof(timestamp), formatSpecifier, localTime) > 0) {
+    return {formatSpecifier};
+  }
+
+  return {};
+#else
+  std::ostringstream timestamp;
+  timestamp << std::put_time(localTime, formatSpecifier);
+  return timestamp.str();
+#endif
+}
+
+/// Returns a string to the current timestamped recovery folder
+std::string getOutputPath() {
+
+  auto timestamp = getTimeStamp();
+  auto timestampedPath = getRecoveryFolder().append(timestamp);
+
+  return timestampedPath;
+}
+
+const std::string OUTPUT_PROJ_NAME = "recovery.mantid";
+
+// Config keys
+const std::string SAVING_ENABLED_CONFIG_KEY = "projectRecovery.enabled";
+const std::string SAVING_TIME_KEY = "projectRecovery.secondsBetween";
+const std::string NO_OF_CHECKPOINTS_KEY = "projectRecovery.numberOfCheckpoints";
+
+// Config values
+const bool SAVING_ENABLED =
+    getConfigBool(SAVING_ENABLED_CONFIG_KEY).get_value_or(false);
+const int SAVING_TIME =
+    getConfigValue<int>(SAVING_TIME_KEY).get_value_or(60); // Seconds
+const int NO_OF_CHECKPOINTS =
+    getConfigValue<int>(NO_OF_CHECKPOINTS_KEY).get_value_or(5);
+
+// Implementation variables
+Mantid::Kernel::Logger g_log("Project Recovery Thread");
+const std::chrono::seconds TIME_BETWEEN_SAVING(SAVING_TIME);
+
+} // namespace
+
+namespace MantidQt {
```

Comment:
Usually namespaces are good practice but as this is currently only part of MantidPlot so can we remove the namespaces as our standard has been that `MantidQt::API` would be part of the `MantidQtAPI` library.